### PR TITLE
fix: Fixed datepicker button overflow

### PR DESCRIPTION
**Pull request by SaiShan**
Closes SAP/fundamental-styles#65

Fixed datepicker button overflow as per issue #65. 
See screenshot. 

Screen Shot 2019-06-13 at 10 39 28 AM

### CODE REVIEW

**Comment by SaiShan** [comment]
> Can we have before and after screenshots for easier review please? Thank you!

Yes! Updated the PR description.

**Comment by jbadan** [comment]
@greg-a-smith do these changes align with your self contained styles plan for buttons?

**Comment by greg-a-smith** [comment]
@SaiShan This is certainly better than before, however, the focus/hover display seems off.  See the attached gif showing focus and hover states between the input and the button.
fd-react-datepicker

**Comment by LeoT7508** [comment]
We need to make updates to the selected state for date picker and  dropdown menu. I will update the design sheet and  share here soon.

File: scss/components/calendar.scss [hunk 1]
```diff
@@ -150,6 +150,8 @@ $fd-calendar-table-width: $fd-calendar-table-item-width * 7 !default;
     height: fd-space(13);
     display: flex;
     align-items: center;
+    width: 276px;
+    margin-left: -12px;
```
Comment (greg-a-smith) [inline scss/components/calendar.scss]: This _looks_ wrong.  There shouldn't be any reason to force a specific pixel width on the header within a calendar.

File: scss/components/calendar.scss [hunk 2]
```diff
@@ -166,6 +168,7 @@ $fd-calendar-table-width: $fd-calendar-table-item-width * 7 !default;
     flex: 1;
     justify-content: space-between;
     align-items: center;
+    padding: 12px;
```
Comment (greg-a-smith) [inline scss/components/calendar.scss]: _Seems_ wrong.  If padding is desired within the popover, it should be the popover that contains the padding (or a child of it that is a parent to the entire calendar component).

File: scss/components/input-group.scss [hunk 3]
```diff
@@ -135,7 +135,13 @@ $block: #{$fd-namespace}-input-group;
         &--button {
           padding: 0;
           [class*="fd-button"] {
```
Comment (greg-a-smith) [inline scss/components/input-group.scss]: You don't need to write attribute-matching selectors for the `class` attribute.  You can just use `&.fd-button`.

File: scss/components/input-group.scss [hunk 4]
```diff
@@ -135,7 +135,13 @@ $block: #{$fd-namespace}-input-group;
...
-            border-radius: 0;
+            border-top-right-radius: 3px;
+            border-bottom-right-radius: 3px;
```
Comment (greg-a-smith) [inline scss/components/input-group.scss]: Is there a variable for this radius?  Seems like it would be used in a bunch of places.

File: scss/components/button.scss [hunk 5]
```diff
@@ -76,7 +76,11 @@ a.#{$block} {
       }
       &:empty {
         padding: 0;
-        &::before,
+        &::before {
+          margin-right: 0;
+          margin-left: 0;
+          margin-bottom: 1px;
+        }
```
Comment (greg-a-smith) [inline scss/components/button.scss]: I'm curious ... when would a button be "empty"?

File: scss/components/dropdown.scss [hunk 6]
```diff
@@ -130,7 +130,7 @@ $block: #{$fd-namespace}-dropdown;
             @if $fd-support-css-var-fallback {
               box-shadow: 0 0 0 1px fd-color-state("hover", "action");
             }
-            box-shadow: 0 0 0 1px var(--fd-color-action-hover);
+            box-shadow: none;
```
Comment (greg-a-smith) [inline scss/components/dropdown.scss]: This is getting rid of the box shadow on the button itself when its popover is displayed, right?

**Comment by greg-a-smith** [comment]
@SaiShan On a separate note, the playground nunjucks templates are emitting `<input type="date">` for the Date Picker component which is then changing the control in the browser.  See below:
Screen Shot 2019-06-24 at 11 18 00 AM

File: scss/components/dropdown.scss (hunk 6 above)
Comment (SaiShan) [reply scss/components/dropdown.scss]: @greg-a-smith Yes, exactly.

File: scss/components/calendar.scss (hunk 1 above)
Comment (SaiShan) [reply scss/components/calendar.scss]: @greg-a-smith Appropriate changes will be made.

File: scss/components/calendar.scss (hunk 2 above)
Comment (SaiShan) [reply scss/components/calendar.scss]: @greg-a-smith Many components are using the popover. Applying padding to the popover would lead to undesired results to the other components using it.

File: scss/components/input-group.scss (hunk 4 above)
Comment (SaiShan) [reply scss/components/input-group.scss]: @greg-a-smith Yes, you are right. Changes will be made.

File: scss/components/button.scss (hunk 5 above)
Comment (SaiShan) [reply scss/components/button.scss]: @greg-a-smith  An empty button is when there is no text between opening and closing tags. 

ie: 
Normal button: `<button class="fd-button--emphasized sap-icon--cart">Add to Cart</button>`
Empty button: `<button class="fd-button sap-icon--cart"></button>`

File: scss/components/calendar.scss (hunk 2 above)
Comment (greg-a-smith) [reply scss/components/calendar.scss]: I agree, it shouldn't be default padding on the popover itself, but some sort of calendar wrapper when it is using within the popover to get the appropriate padding that wouldn't exist on a standalone calendar.

File: scss/components/input-group.scss (hunk 3 above)
Comment (SaiShan) [reply scss/components/input-group.scss]: @greg-a-smith The proper styling is not applied when using `&.fd-button` as you can see on the following screenshot (incorrect border radius). 

Screen Shot 2019-06-25 at 4 00 20 PM

File: scss/components/calendar.scss (hunk 2 above)
Comment (SaiShan) [reply scss/components/calendar.scss]: I believe the fd-calendar is the wrapper which contains all the elements of a calendar. In the newest version of this file, the padding are used match the latest design sheet.

File: docs/pages/components/calendar.md [hunk 7]
```diff
@@ -175,225 +175,227 @@ These are dynamic classes which are affected by user selections
 ## Example: Dates, Months, Years
 
 {% capture default-calendar-dates-months-years %}
-<div class="fd-calendar">
-	<header class="fd-calendar__header">
-	<div class="fd-calendar__navigation">
-		<div class="fd-calendar__action">
-			<button class="fd-button--standard fd-button--light fd-button--compact sap-icon--slim-arrow-left" aria-label="Previous" aria-disabled="false"></button>
+<div style="width: 292px">
```
Comment (greg-a-smith) [inline docs/pages/components/calendar.md]: Setting a fixed pixel width here _seems_ wrong.

File: docs/pages/components/combobox-input.md [hunk 8]
```diff
@@ -45,7 +45,7 @@ summary:
               <div class="fd-input-group fd-input-group--after fd-input-group--compact">
                   <input type="text" class="fd-input fd-input--compact" id="" placeholder="Select Fruit">
                   <span class="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button">
-                      <button class=" fd-button--light sap-icon--navigation-down-arrow"></button>
+                      <button class=" fd-button--light fd-button--compact sap-icon--navigation-down-arrow"></button>
```
Comment (greg-a-smith) [inline docs/pages/components/combobox-input.md]: Although probably a necessary change, it completely unrelated to the DatePicker button.  This should be reverted and put into its own PR.

File: docs/pages/components/date-picker.md [hunk 9]
```diff
@@ -22,18 +22,18 @@ This component mostly relies on the CSS of other components and has very little
...
 			<div class="fd-input-group fd-input-group--after">
-				<input type="text" value="" placeholder="Pick a date">
+				<input type="text" class="fd-input" value="" placeholder="Pick a date">
 				<span class="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button">
 				<button class="fd-popover__control fd-button--light sap-icon--calendar" aria-controls="Itcgq828" aria-expanded="false" aria-haspopup="true" aria-hidden="true"></button>
 				</span>
 			</div>
 		</div>
-		<div class="fd-popover__body fd-popover__body--right fd-popover__body--no-arrow" aria-hidden="true" id="Itcgq828">
+		<div class="fd-popover__body fd-popover__body--left fd-popover__body--no-arrow" aria-hidden="true" id="Itcgq828">
```
Comment (greg-a-smith) [inline docs/pages/components/date-picker.md]: Is this right?  Is it really moving from right-aligned to left-aligned?

File: docs/pages/components/date-picker.md [hunk 10]
```diff
@@ -207,16 +207,16 @@ This component mostly relies on the CSS of other components and has very little
 			<div class="fd-input-group fd-input-group--after fd-input-group--compact">
 				<input type="text" value="" placeholder="Pick a date" class="fd-input fd-input--compact">
 				<span class="fd-input-group__addon fd-input-group__addon--after fd-input-group__addon--button">
-				<button class="fd-popover__control fd-button--light sap-icon--calendar" aria-controls="Itcgq82" aria-expanded="false" aria-haspopup="true" aria-hidden="true"></button>
+				<button class="fd-popover__control fd-button--light fd-button--compact sap-icon--calendar" aria-controls="Itcgq82" aria-expanded="false" aria-haspopup="true" aria-hidden="true"></button>
 				</span>
 			</div>
 		</div>
-		<div class="fd-popover__body fd-popover__body--right fd-popover__body--no-arrow" aria-hidden="true" id="Itcgq82">
+		<div class="fd-popover__body fd-popover__body--left fd-popover__body--no-arrow" aria-hidden="true" id="Itcgq82">
```
Comment (greg-a-smith) [inline docs/pages/components/date-picker.md]: Is this right? Is it really moving from right-aligned to left-aligned?

File: scss/components/button.scss [hunk 11]
```diff
@@ -324,6 +331,20 @@ a.#{$block} {
         @include fd-var-color("background-color", fd-color("status", 2), --fd-button-background-color);
       }
     }
+    &[aria-expanded="true"] {
```
Comment (greg-a-smith) [inline scss/components/button.scss]: We shouldn't be using `aria-*` attributes for CSS selectors.

File: scss/components/input-group.scss (hunk 3 above)
Comment (greg-a-smith) [reply scss/components/input-group.scss]: Okay, you can leave this for now.  I have created #138 to go through and refactor blocks like this.

File: scss/components/calendar.scss [hunk 12]
```diff
@@ -160,12 +153,16 @@ $fd-calendar-table-width: $fd-calendar-table-item-width * 7 !default;
   }
   &__content {
     padding: fd-space(xs) 0;
+    padding-left: 12px;
+    padding-right: 12px;
```
Comment (greg-a-smith) [inline scss/components/calendar.scss]: These three lines can be rolled up into the single line: `padding: fd-space(xs) 12px;`

File: scss/components/button.scss [hunk 13]
```diff
@@ -230,6 +235,8 @@ a.#{$block} {
       --fd-button-color: var(--fd-color-action-1);
       @include fd-var-color("color", fd-color("action", 1), --fd-button-color);
       @include fd-var-color("background-color", fd-color("neutral", 1), --fd-button-background-color);
+      box-shadow: 0 0 0 1px var(--fd-color-action-hover,#085CAF);
+      z-index: 1;
```
Comment (greg-a-smith) [inline scss/components/button.scss]: I wouldn't have expected all these changes in the `button.scss` file for a DatePicker / InputGroup button issue.  Are these truly needed and if so, should they be part a PR to change or refactor buttons?

File: docs/pages/components/date-picker.md (hunk 9 above)
Comment (SaiShan) [reply docs/pages/components/date-picker.md]: According to the newest design sheet, it should be left-aligned.

**Comment by SaiShan** [comment]
As agreed with @greg-a-smith, Will close this PR and reopen a new one with the changes concerning only the button overflow. For the combobox and styling of button and calendar, separate issues will be created.
